### PR TITLE
SDK - Compiler - Make it possible to create more portable pipelines

### PR DESCRIPTION
**Pull request by Ark-kun**
This change allows directly passing the PipelineConf instance to compiler or launcher which makes it easier to create portable pipelines by allowing the environment-specific configuration to be directly passed to the environment-specific launcher.

Background:
PipelineConf holds all pipeline-level configuration including `op_transformers`, `image_pull_secrets` etc. Some of these are specific to particular execution environment (e.g. GCP secret or Argo artifact location or Kubernetes-specific options).
Previously, the only way to modify `PipelineConf` was to do it inside the piepline function. That tied the pipeline function to specific execution environment (e.g. GCP, Argo or Kubernetes)

Solution: This change allows directly passing the PipelineConf instance to compiler or launcher. This allows writing portable enlauncher and environment agnostic pipeline functions. All environment-specific configurations can be moved to launching stage.

Before:
```python
# Defining pipeline
def my_pipeline():
    # portable pipeline code

    dsl.get_pipeline_conf().add_op_transformer(gcp.use_gcp_secret('user-gcp-sa'))

# Launching pipeline
kfp.Clinet().create_run_from_pipeline_func(my_pipeline, arguments={})
```

After:
```python
# Defining pipeline
def my_pipeline():
    # portable pipeline code

# Launching pipeline
pipeline_conf = dsl.PipelineConf()
pipeline_conf.add_op_transformer(gcp.use_gcp_secret('user-gcp-sa'))
kfp.Clinet().create_run_from_pipeline_func(my_pipeline, arguments={}, pipeline_conf=pipeline_conf)
```

After 2 *(launching same portable pipeline using different launchers):
```python
# Loading portable pipeline
from portable_pipeline import my_pipeline

# Launching pipeline on Kubeflow
pipeline_conf = dsl.PipelineConf()
pipeline_conf.add_op_transformer(gcp.use_gcp_secret('user-gcp-sa'))
kfp.Clinet().create_run_from_pipeline_func(my_pipeline, arguments={}, pipeline_conf=pipeline_conf)

# Launching pipeline on locally (not implemented yet)
kfp.run_pipeline_func_locally(my_pipeline, arguments={})
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2271)

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -608,7 +608,7 @@ def _create_volumes(self, pipeline):
     volumes.sort(key=lambda x: x['name'])
     return volumes
 
-  def _create_pipeline_workflow(self, args, pipeline, op_transformers=None):
+  def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeline_conf=None):
```
Comment (numerology) [inline sdk/python/kfp/compiler/compiler.py]: What will happen if a user set op transformer in op_transformers and also in pipeline_conf?

Comment (numerology) [reply sdk/python/kfp/compiler/compiler.py]: This api is hidden, but is it possible for a user to call dsl.get_pipeline_conf().add_op_transformer(gcp.use_gcp_secret('user-gcp-sa')) inside the pipline definition, and then do something not compatible in pipeline_conf?

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -778,8 +781,10 @@ def create_workflow(self,
     with dsl.Pipeline(pipeline_name) as dsl_pipeline:
       pipeline_func(*args_list)
 
+    pipeline_conf = pipeline_conf or dsl_pipeline.conf # Configuration passed to the compiler is overriding. Unfortunately, it's not trivial to detect whether the dsl_pipeline.conf was ever modified.
```
Comment (numerology) [inline sdk/python/kfp/compiler/compiler.py]: According to this line, shall we add some doc saying that the one set in pipeline_conf shadows that defined by dsl.get_pipline_conf ?

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/compiler/compiler.py]: Yes this is possible.
In this case the configuration explicitly specified by the "pipeline run submitter" overrides possible configuration set by the "pipeline author".

File: sdk/python/kfp/compiler/compiler.py (hunk 2 above)
Comment (Ark-kun) [reply sdk/python/kfp/compiler/compiler.py]: I've added docstring which documents the overriding behavior.

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/compiler/compiler.py]: Added docstring describing the behavior.

**Pull request merged**